### PR TITLE
test: change the UDP multicast to a administratively scoped range

### PR DESCRIPTION
**Pull request by juanarbol**
Fixes: https://github.com/libuv/libuv/issues/4263

### CODE REVIEW

File: test/test-udp-multicast-join.c [hunk 1]
```diff
@@ -29,7 +29,14 @@
...
   ASSERT_NE((uv_udp_t*)(handle) == &server || (uv_udp_t*)(handle) == &client, 0)
 
+/* Use 224.0.0.x subnet for MAC OS 13.0 and above
+ * Refs https://github.com/libuv/libuv/issues/4263
+ */
+#if defined(__APPLE__) && defined(MAC_OS_VERSION_13_0)
+#define MULTICAST_ADDR "224.0.0.9"
+#else
 #define MULTICAST_ADDR "239.255.0.1"
+#endif
```
[the remark this answers is not in the document]
Comment (santigimeno) [reply test/test-udp-multicast-join.c:39]: I don't mind changing it though the original issue doesn't happen in my `macOS 14.3`

**Comment by juanarbol** [comment]
After a lot of pain and horrible macOS documentation, I did not found the actual reason, but I created a a minimal repro case for this issue, and it is persistent, even without libuv.

```c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <sys/errno.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/ip.h>
#include <net/if.h>

#define MULTICAST_ADDR "239.255.0.1"
#define PORT 9123

int main() {
    int sockfd;
    struct sockaddr_in addr;
    struct ip_mreq mreq;
    struct ifreq ifr;

    // Create a UDP socket
    if ((sockfd = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
        perror("socket creation failed");
        exit(EXIT_FAILURE);
    }

    // Set up the multicast address
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_ANY);
    addr.sin_port = htons(PORT);

    // Bind the socket
    if (bind(sockfd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
        perror("bind failed");
        exit(EXIT_FAILURE);
    }

    // Set up multicast group membership
    memset(&mreq, 0, sizeof(mreq));
    mreq.imr_multiaddr.s_addr = inet_addr(MULTICAST_ADDR);
    mreq.imr_interface.s_addr = INADDR_ANY;
    if (setsockopt(sockfd, IPPROTO_IP, IP_ADD_MEMBERSHIP, &mreq, sizeof(mreq)) < 0) {
        perror("setsockopt failed");
        exit(errno);
    }

    printf("Multicast group joined successfully\n");

    // Close the socket
    close(sockfd);

    return 0;
}
```

That prints `setsockopt failed: Exec format error`, other IP address like `235.0.0.0` (which is a reserved one) behaves more weird. It passes for the first time, then it fails with -16 errno, then -8 forever. Sometimes, within the range `235.0.0.0` to `238.255.255.255` first fails with -16, then -8, then it just always passes.

```sh
➜ ./build/uv_run_tests_a udp_multicast_join udp_multicast_join
Assertion failed in /Users/juanjose/GitHub/libuv/test/test-udp-multicast-join.c on line 131: `r == 0` (-16 == 0)
[1]    56998 abort      ./build/uv_run_tests_a udp_multicast_join udp_multicast_join

libuv on  juan/fix-macos-test [⇡!]
➜ ./build/uv_run_tests_a udp_multicast_join udp_multicast_join
setsockopt failed: : Exec format error
Assertion failed in /Users/juanjose/GitHub/libuv/test/test-udp-multicast-join.c on line 163: `r == 0` (-8 == 0)
[1]    57723 abort      ./build/uv_run_tests_a udp_multicast_join udp_multicast_join
```

File: test/test-udp-multicast-join.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (juanarbol) [reply test/test-udp-multicast-join.c]: Addressed in https://github.com/libuv/libuv/pull/4314/commits/9ec71600dd6f1ee80918abe92bc33e19727fc3b8

**Comment by bnoordhuis** [comment]
@juanarbol thanks for investigating. I don't really have suggestions what to do next. Maybe we should just chalk this up as "fact of life" (and: "closed-source proprietary operating systems suck")

**Comment by juanarbol** [comment]
> @juanarbol thanks for investigating. I don't really have suggestions what to do next. Maybe we should just chalk this up as "fact of life" (and: "closed-source proprietary operating systems suck")

I found [the solution](https://stackoverflow.com/a/65537860), you were completely right since the beginning.

Screenshot 2024-05-09 at 1 52 19

After those settings, I get the regular "Allow this..." and the test never fails in `v1.x`

**Pull request closed without merging**